modules/serverless-apiserversource.adoc

```diff
@@ -0,0 +1,231 @@
+// Module included in the following assemblies:
+//
+// serverless/knative_eventing/serverless-kn-source.adoc
+
+[id="serverless-apiserversource_context"]
+= Using the ApiServerSource
+
+ApiServerSource is an event source that can be used to connect an event sink, such as a Knative service, to the Kubernetes API server.
+ApiServerSource watches for Kubernetes events and forwards them to the Knative Eventing broker.
+
+.Prerequisites
+
+* Using ApiServerSource requires you to have installed Knative Serving and Eventing. For more information, see the documentation on xref:../../serverless/installing-openshift-serverless.adoc#installing-openshift-serverless[Installing OpenShift Serverless].
+
+* You must have created the `default` broker in the same namespace as the one defined in the ApiServerSource YAML. For more information, see the documentation on xref:../../serverless/knative_eventing/serverless-using-brokers.adoc#serverless-using-brokers[Using brokers with Knative Eventing].
+
+* For steps that use `kn`, you will need to have the `kn` CLI installed. For more information, see the documentation on xref:../../serverless/installing-kn.adoc#installing-kn[Installing the Knative CLI].
+
+.Procedure
+
+. Create a Service Account for the ApiServerSource.
++
+You can do this by creating a file named `serviceaccount.yaml` and copying the following sample code into it:
++
+----
+apiVersion: v1
+kind: ServiceAccount
+metadata:
+  name: events-sa
+  namespace: default
+
+---
+apiVersion: rbac.authorization.k8s.io/v1
+kind: ClusterRole
+metadata:
+  name: event-watcher
+rules:
+  - apiGroups:
+      - ""
+    resources:
+      - events
+    verbs:
+      - get
+      - list
+      - watch
+
+---
+apiVersion: rbac.authorization.k8s.io/v1
+kind: ClusterRoleBinding
+metadata:
+  name: k8s-ra-event-watcher
+roleRef:
+  apiGroup: rbac.authorization.k8s.io
+  kind: ClusterRole
+  name: event-watcher
+subjects:
+  - kind: ServiceAccount
+    name: events-sa
+    namespace: default
+----
++
+[NOTE]
+====
+If you want to re-use an existing Service Account with the appropriate permissions, you must modify the `serviceaccount.yaml` for that Service Account.
+====
++
+Create the service account by entering the following command:
++
+----
+$ oc apply --filename serviceaccount.yaml
+----
+
+. Create an ApiServerSource event source.
++
+You can do this by creating a file named `k8s-events.yaml` and copying the following sample code into it:
++
+----
+apiVersion: sources.knative.dev/v1alpha1
+kind: ApiServerSource
+metadata:
+  name: testevents
+  namespace: default
+spec:
+  serviceAccountName: events-sa
+  mode: Resource
+  resources:
+    - apiVersion: v1
+      kind: Event
+  sink:
+    ref:
+      apiVersion: eventing.knative.dev/v1beta1
+      kind: Broker
+      name: default
+----
++
+[NOTE]
+====
+If you want to consume events from a different namespace or use a different Service Account, you need to modify the sample `k8s-events.yaml` file accordingly.
```

Comment:
I think this instruction is to vague, as it requires multiple steps to consume events from a different namespace (let's say "otherns"):

* You need to create a service-account in "otherns" and a new clusterolebinding for this other service-account 
* You need to change the namespace of the ApiServerSource to "otherns" (in `metadata:` of the ApiServerSource)
* You need to add a `namespace: default` to the broker defined in the `ref:` part of the ApiServerSource.

The whole yaml you would need to add would be (in addition to the steps already done)

```yaml
apiVersion: v1
kind: ServiceAccount
metadata:
  # The name can be reused here because its in another namespace
  name: events-sa
  namespace: otherns
---
apiVersion: rbac.authorization.k8s.io/v1
kind: ClusterRoleBinding
metadata:
  # Note the different name that is required
  name: k8s-ra-event-watcher-otherns
roleRef:
  apiGroup: rbac.authorization.k8s.io
  kind: ClusterRole
  # This is the role that already has been defined
  name: event-watcher
subjects:
  - kind: ServiceAccount
    name: events-sa
    # New namespace here
    namespace: otherns
```

```yaml
apiVersion: sources.knative.dev/v1alpha1
kind: ApiServerSource
metadata:
  name: testevents
  # The namespace here is important and points to the other namespace
  namespace: otherns
spec:
  # This is the name of the service account just created above
  serviceAccountName: events-sa
  mode: Resource
  resources:
    - apiVersion: v1
      kind: Event
  sink:
    ref:
      apiVersion: eventing.knative.dev/v1alpha1
      kind: Broker
      name: default
      # Here you need to point to the broker in the original namespace, default in this example
      namespace: default
```

Reply:
@matzew is it really recommended to send events to brokers in another namespace ? It works (just have verified), but is this the canonical way to get events from other namespaces (i.e. creating a second ApiServerSource)

Reply:
@abrennan89 in short, I probably would leave out that note about how to get to events from other namespaces ;-)